### PR TITLE
Feat(FN-420): function-number-generator in Bicep

### PR DESCRIPTION
**Pull request by OxonianCambion**
### Introduction

We need to be able to manage the infrastructure in declarative IaC - we've chosen Bicep.
The first phase is to be able to:

- produce a complete new deployment environment (feature) mirroring the current ones.
- be in a position to take over the extant environments.
- Note that there have been some manual changes to the environments. This work also functions as an audit of the current infrastructure.

Possible enhancements / updates will be noted but not implemented in the first phase.

This card covers 
* Implementing the function-number-generator Azure Function infrastructure in Bicep
### Resolution
* The function-number-generator Azure Function infrastructure is implemented in Bicep
* Environment variables have been checked and categorised

### CODE REVIEW

File: infrastructure/resource_group_level/modules/function-number-generator.bicep [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+param storageAccountName string
+
+
+// These values are taken from GitHub secrets injected in the GHA Action
+@secure()
+param secureSettings object = {
+  APIM_TFS_KEY: 'test-value'
+  APIM_TFS_VALUE: 'test-value'
+  APIM_TFS_URL: 'test-value'
+  APIM_MDM_KEY: 'test-value'
```
Comment (OxonianCambion) [inline infrastructure/resource_group_level/modules/function-number-generator.bicep:16]: Note that the `APIM_MDM_*` keys were not in the CLI script, but should be. @abhi-markan will add them in his own PR.

Comment (IgnatG) [reply infrastructure/resource_group_level/modules/function-number-generator.bicep:16]: this should be added now

File: infrastructure/resource_group_level/modules/function-number-generator.bicep [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+var additionalSettings = {
+  APPINSIGHTS_INSTRUMENTATIONKEY: applicationInsights.properties.InstrumentationKey
+  AzureWebJobsStorage: 'DefaultEndpointsProtocol=https;AccountName=${storageAccountName};AccountKey=${storageAccountKey}'
+  // TODO:FN-684 DOCKER_CUSTOM_IMAGE_NAME is overridden by linuxFxVersion. Remove if not necessary.
+  DOCKER_CUSTOM_IMAGE_NAME: dockerImageName
+  DOCKER_ENABLE_CI: 'true'
+  DOCKER_REGISTRY_SERVER_URL: '${containerRegistryName}.azurecr.io'
+  DOCKER_REGISTRY_SERVER_USERNAME: dockerRegistryServerUsername
+  FUNCTION_APP_EDIT_MODE: 'readOnly'
+  FUNCTIONS_EXTENSION_VERSION: '~3'
```
Comment (IgnatG) [inline infrastructure/resource_group_level/modules/function-number-generator.bicep:60]: fine with v3. but we would need to upgrade it at some point to the latest version. I think it's 4?

File: infrastructure/resource_group_level/modules/function-number-generator.bicep [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+      acrUseManagedIdentityCreds: false
+      alwaysOn: true
+      http20Enabled: true
+      functionAppScaleLimit: 0
+      minimumElasticInstanceCount: 1
+      // The following Fields have been added after comparing the generated insance export with dev
+      vnetRouteAllEnabled: true
+      // Note that the following only appear in the separate config object on export, but we can set them inline.
+      ftpsState: 'Disabled'
+      scmMinTlsVersion: '1.0'
```
Comment (IgnatG) [inline infrastructure/resource_group_level/modules/function-number-generator.bicep:102]: should this be `1.2`?

**Review by IgnatG — COMMENTED** [review]
It looks good overall. the main question is around minimum TLS version

File: infrastructure/resource_group_level/modules/function-number-generator.bicep (hunk 2 above)
Comment (OxonianCambion) [reply infrastructure/resource_group_level/modules/function-number-generator.bicep:60]: Yes - this is tracked in FN-696, in the https://ukef-dtfs.atlassian.net/browse/FN-692 Epic (Infrastructure Enhancements)

**Pull request merged**